File: measurements/stability_measurement.py
"""Port of APS2/Measurements/StabilityMeasurement.m."""

from __future__ import annotations
from dataclasses import dataclass, asdict
from .base_measurement import BaseMeasurement


@dataclass
class StabIV:
    startV: float = 0.0
    maxV: float = 0.2
    minV: float = -0.2
    points: int = 1001
    sweep_dir: str = "up"
    scanrate: int = 450000
    process_number: int = 1


@dataclass
class StabGate:
    initV: float = 0.0
    minV: float = -0.5
    maxV: float = 0.5
    dV: float = 0.001
    ramp_rate: float = 0.5
    process: str = "Fixed_AO"
    process_number: int = 3
    waiting_time: float = 0.1
    sweep_dir: str = "up"


class StabilityMeasurement(BaseMeasurement):
    MATLAB_CLASS = "StabilityMeasurement"

    def __init__(self, bridge, settings: dict,
                 iv: StabIV | None = None, gate: StabGate | None = None):
        super().__init__(bridge, settings)
        self.iv = iv or StabIV()
        self.gate = gate or StabGate()

    def create(self):
        super().create()
        self.bridge.assign_fields(self._var + ".IV",   asdict(self.iv))
        self.bridge.assign_fields(self._var + ".Gate", asdict(self.gate))

    def stop(self):
        if self._created:
            self.bridge.eval(f"{self._var}.Stop();")

File: APS2/Measurements/StabilityMeasurement.m
classdef StabilityMeasurement < BaseMeasurement
    % StabilityMeasurement - Performs stability measurements over time/gate
    
    properties
        IV struct
        Gate struct
        StopFlag (1,1) logical = false
    end
    
    methods
        function obj = StabilityMeasurement(settings)
            obj@BaseMeasurement(settings);
            obj.Settings.type = 'Stability';
            
            % IV Params
            obj.IV.startV = 0; 
            obj.IV.maxV = 0.2;
            obj.IV.minV = -0.2;
            obj.IV.points = 1001;
            obj.IV.sweep_dir = 'up';
            obj.IV.scanrate = 450000;
            obj.IV.process_number = 1;
            
            % Gate Params
            obj.Gate.initV = 0;
            obj.Gate.minV = -0.5;
            obj.Gate.maxV = 0.5;
            obj.Gate.dV = 0.001;
            obj.Gate.ramp_rate = 0.5;
            obj.Gate.process = 'Fixed_AO';
            obj.Gate.process_number = 3;
            obj.Gate.waiting_time = 0.1;
        end
        
        function Run(obj)
            fprintf('--- Starting Stability Measurement ---\n');
            
            % 1. Init ADwin
            obj.InitADwinProcess(obj.IV);
            obj.InitADwinProcess(obj.Gate);
            
            % 2. Generate Gate Vector
            gateV = obj.Gate.minV : obj.Gate.dV : obj.Gate.maxV;
            if strcmp(obj.Gate.sweep_dir, 'down')
                gateV = fliplr(gateV);
            end
            
            obj.IV.repeat = length(gateV);
            
            % 3. Measurement Loop
            for i = 1:obj.IV.repeat
                if obj.StopFlag, break; end
                
                % Set Gate
                obj.Gate.setV = gateV(i);
                % Apply_fixed_voltage(obj.Settings, obj.Gate); % (Assuming global function)
                
                pause(obj.Gate.waiting_time);
                
                % Run IV
                try
                    obj.IV.index = i;
                    obj.IV = Run_sweep(obj.Settings, obj.IV); % (Assuming global function)
                    
                    % Store Data
                    obj.Data.IV = obj.IV; 
                    obj.Data.Gate = obj.Gate;
                    
                    % Realtime Plot
                    % Realtime_sweep3D(...) 
                catch ME
                    fprintf(2, 'Stability Error: %s\n', ME.message);
                end
            end
            
            obj.Save();
        end
        
        function Stop(obj)
            obj.StopFlag = true;
        end
    end
end